libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,205 @@
+            if (_offs >= _len) {
+                _state = State::GET_CHKSUML;
+            }
+            break;
+
+        case State::GET_CHKSUML:
+            _lchksum = c;
+            _state = State::GET_CHKSUMH;
+            break;
+
+        case State::GET_CHKSUMH:
+            // Validate checksum
+            if (_chksum == (c << 8) + _lchksum) {
+                handleTelemetryMessage();
+            }
+            break;
+        case State::DISCARD:
+            _state = State::GET_LENGTH;
+            break;
+        }
+    }
+}
+
+void AP_IBus_Telem::handleTelemetryMessage()
+{
+    // Checksum is all fine Execute command -
+    const uint8_t adr = _buffer[0] & 0x0f;
+
+    // all sensor data commands go here
+    // we only process the len==1 commands (=message length is 4 bytes incl overhead) to prevent the case the
+    // return messages from the UART TX port loop back to the RX port and are processed again. This is extra
+    // precaution as it will also be prevented by the PROTOCOL_TIMEGAP required
+    if (adr > sizeof(_sensors) || adr == 0 || _len != 1) {
+        return;
+    }
+
+    _sensorinfo *s = &_sensors[adr - 1];
+
+    switch (_buffer[0] & 0x0f0) {
+    case PROTOCOL_COMMAND_DISCOVER: { // 0x80, discover sensor
+        // echo discover command: 0x04, 0x81, 0x7A, 0xFF
+        struct protocol_command_discover_response_t {
+            uint8_t command_length;
+            uint8_t address;
+            uint16_t checksum;
+        } packet {
+            PROTOCOL_FOUR_LENGTH,
+            (uint8_t)(PROTOCOL_COMMAND_DISCOVER | adr)
+        };
+        populate_checksum((uint8_t*)&packet, sizeof(packet));
+        _port->write((uint8_t*)&packet, sizeof(packet));
+        break;
+    }
+    case PROTOCOL_COMMAND_TYPE: { // 0x90, send sensor type
+        // echo sensortype command: 0x06 0x91 0x00 0x02 0x66 0xFF
+        struct protocol_command_type_response_t {
+            uint8_t command_length;
+            uint8_t address;
+            uint8_t type;
+            uint8_t length;
+            uint16_t checksum;
+        } packet {
+            PROTOCOL_SIX_LENGTH,
+            (uint8_t)(PROTOCOL_COMMAND_TYPE | adr),
+            s->sensor_type,
+            s->sensor_length
+        };
+        populate_checksum((uint8_t*)&packet, sizeof(packet));
+        _port->write((uint8_t*)&packet, sizeof(packet));
+        break;
+    }
+    case PROTOCOL_COMMAND_VALUE: { // 0xA0, send sensor data
+        struct protocol_command_2byte_value_response_t {
+            uint8_t command_length;
+            uint8_t address;
+            uint16_t value;
+            uint16_t checksum;
+        } packet {
+            PROTOCOL_SIX_LENGTH,
+            (uint8_t)(PROTOCOL_COMMAND_VALUE | adr),
+            (uint16_t)s->sensor_value
+        };
+        populate_checksum((uint8_t*)&packet, sizeof(packet));
+        _port->write((uint8_t*)&packet, sizeof(packet));
+        break;
+    }
+    default:
+        break;
```

Comment:
Perhaps add a comment explaining its an unrecognised command
```suggestion
        // unrecognised command - ignore it
        break;
```